README.md
# NAME

Catalyst::View::Xslate - Text::Xslate View Class

# SYNOPSIS

    package MyApp::View::Xslate;
    use Moose;
    extends 'Catalyst::View::Xslate';

    1;

# VIEW CONFIGURATION

You may specify the following configuration items in from your config file
or directly on the view object.

## catalyst\_var

The name used to refer to the Catalyst app object in the template

## template\_extension

The suffix used to auto generate the template name from the action name
(when you do not explicitly specify the template filename);

Do not confuse this with the `suffix` option, which is passed directly to
the Text::Xslate object instance. This option works on the filename used
for the initial request, while `suffix` controls what `cascade` and
`include` directives do inside Text::Xslate.

## content\_charset

The charset used to output the response body. The value defaults to 'UTF-8'.

## encode\_body

By default, output will be encoded to `content_charset`.
You can set it to 0 to disable this behavior.
(you need to do this if you're using `Catalyst::Plugin::Unicode::Encoding`)

__NOTE__ Starting with [Catalyst](https://metacpan.org/pod/Catalyst) version 5.90080 Catalyst will automatically
encode to UTF8 any text like body responses.  You should either turn off the
body encoding step in this view using this attribute OR disable this feature
in the application (your subclass of Catalyst.pm).

    MyApp->config(encoding => undef);

Failure to do so will result in double encoding.

## Text::Xslate CONFIGURATION

The following parameters are passed to the Text::Xslate constructor.
When reset during the life cyle of the Catalyst app, these parameters will
cause the previously created underlying Text::Xslate object to be cleared

## path

## cache\_dir

## cache

## header

## escape

## type

## footer

## function

## input\_layer

## module

## syntax

## verbose

## suffix

Use this to enable TT2 compatible variable methods via Text::Xslate::Bridge::TT2 or Text::Xslate::Bridge::TT2Like

    package MyApp::View::Xslate;
    use Moose;
    extends 'Catalyst::View::Xslate';

    has '+module' => (
        default => sub { [ 'Text::Xslate::Bridge::TT2Like' ] }
    );

# preload

Boolean flag indicating if templates should be preloaded. By default this is enabled.

Preloading templates will basically cutdown the cost of template compilation for the first hit.

## expose\_methods

Use this option to specify methods from the View object to be exposed in the
template. For example, if you have the following View:

    package MyApp::View::Xslate;
    use Moose;
    extends 'Catalyst::View::Xslate';

    sub foo {
        my ( $self, $c, @args ) = @_;
        return ...; # do something with $self, $c, @args
    }

then by setting expose\_methods, you will be able to use $foo() as a function in
the template:

    <: $foo("a", "b", "c") # calls $view->foo( $c, "a", "b", "c" ) :>

`expose_methods` takes either a list of method names to expose, or a hash reference, in order to alias it differently in the template.

    MyApp::View::Xslate->new(
        # exposes foo(), bar(), baz() in the template
        expose_methods => [ qw(foo bar baz) ]
    );

    MyApp::View::Xslate->new(
        # exposes $foo_alias(), $bar_alias(), $baz_alias() in the template,
        # but they will in turn call foo(), bar(), baz(), on the view object.
        expose_methods => {
            foo => "foo_alias",
            bar => "bar_alias",
            baz => "baz_alias",
        }
    );

NOTE: you can mangle the process of building the exposed methods, see `build_exposed_method`.

# METHODS

# `$view-`process($c)>

Called by Catalyst.

## `$view-`render($c, $template, \\%vars)>

Renders the given `$template` using variables \\%vars.

`$template` can be a template file name, or a scalar reference to a template
string.

    $view->render($c, "/path/to/a/template.tx", \%vars );

    $view->render($c, \'This is a xslate template!', \%vars );

## `$view-`preload\_templates>

Preloads templates in $view->path.

## `$view-`build\_exposed\_method>

Hook point for mangling the building process of exposed methods.

# AUTHOR

Copyright (c) 2010 Daisuke Maki `<daisuke@endeworks.jp>`

# LICENSE 

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See http://www.perl.com/perl/misc/Artistic.html


Changes
Changes
=======

{{$NEXT}}

0.00019 2015-09-29T23:50:59Z
    * Apply changes from PR #21 as #23. This should fix memory leaks
      introduced in `expose_method` related code (Mark Ellis)

0.00018 2015-01-04T06:28:41Z
    * compatibility shim added to test cases so that the encoding tests will
      pass with Catalyst v 5.90080+
    * Documentation on UTF8 and default encoding on new Catalyst.
    * Updated copyeright notices

0.00017 - 28 Feb 2013
    * Merge pullreq #17: Check passed parameters are defined, not for their
      boolean values (wreis)
    * Merge pullreq #18: Add type attribute (wreis)

0.00016 - 17 Oct 2012
    * Merge pullreq #16: Add a build_expose_method (wreis)

0.00015 - 12 May 2012
    * Merge pullreq #14: Remove unused lazy building (wreis)
    * Merge pullreq #15: Add template preloading (wreis)
    * Add "preload" parameter to toggle preloading on/off

0.00014 - 07 May 2012
    * Merge pullreq #13: Use Catalyst::Component::ApplicationAttribute (wreis)
    * Merge pullreq #12: Change type of 'verbose' option to int (wreis)
    * Merge pullreq #8: Add encode_body attr to avoid automatically
                        encode the body (sunnavy)
                        (Author's note: I recognize that there's a time and
                        place where this is useful, but if you find yourself
                        needing this and you're doing something that doesn't
                        require the entire rendered body to run it, then
                        you're doing something wrong. Instead, see if you can
                        avoid it by using a combination of Text::Xslate's
                        function/module attrs and stuff like html_builder())

0.00013 - 25 Feb 2012
    * Add header/footer support(rt #75256)

0.00012 - 13 Apr 2011
    * Change how 'exposed_methods' are defined. Now functions are generated
      at render time for every request, and is available in the stash
      (John Napiorkowski)

0.00011 - 28 Feb 2011
    * Change TEMPLATE_EXTENSION to template_extension in the helper
    * Factor out the method to generate Xslate object (Getty)

0.00010 - 10 Feb 2011
    * Fix attribute specification for 'cache' to accept Int as
      Text::Xslate accepts values 0, 1, or 2 (travail, gfx)

0.00009 - 19 Jan 2011
    * Added 'expose_methods' configuration option similar to the TT View
      (John Napiorkowski)
    * test for above (John Napiorkowski)
    * allow you to pass a scalar ref to ->render so that you can render
      literal / inlined templates (John Napiorkowski)
    * Added docs

0.00008 - 28 Oct 2010
    * Add content_charset to specify the encoding (hiratara)
        -> You NEED to pass in a flagged UTF-8 to body

0.00007 - 21 Sep 2010
    * Add helpers to generate stubs (garu)
    * Add tests for it (garu, lestrrat)

0.00006 - 15 Jul 2010
    * Update tests to work with newer Text::Xslate
    * Require latest Text::Xslate

0.00005 - 09 Jun 2010
    * Initialize xslate attribute on ACCEPT_CONTEXT

0.00004 - 09 Jun 2010
    * s/functions/function/g

0.00003 - 22 May 2010
    * Fix dependency

0.00002 - 19 May 2010
    * Fix error display

0.00001 - 19 May 2010
    * Initial release


t/lib/TestApp.pm
package TestApp;
use strict;
use warnings;
use Catalyst; # qw/-Debug/;

our $VERSION = '0.00018';

__PACKAGE__->config(
    name                  => 'TestApp',
    default_message       => 'hi',
    default_view          => 'Xslate::Pkgconfig',
    encoding => undef
);

__PACKAGE__->setup;


t/lib/TestApp/Controller/Root.pm
package TestApp::Controller::Root;
use Moose;
use namespace::autoclean;

BEGIN {
    extends 'Catalyst::Controller';
}

__PACKAGE__->config(namespace => '');

sub test_render
    :Local
{
    my ($self, $c) = @_;

    $c->stash->{message} = eval {
        $c->view('Xslate::Appconfig')->render($c, $c->req->param('template'), {param => $c->req->param('param') || ''})
    };
    if (my $err = $@) {
        $c->response->body($err);
        $c->response->status(403);
    } else {
        $c->stash->{template} = 'test.tx';
    }

}

# XXX From View::TT, but not supported in Xslate
# sub test_msg : Local {
#     my ($self, $c) = @_;
#     my $tmpl = $c->req->param('msg');
# 
#     $c->stash->{message} = $c->view('Xslate::AppConfig')->render($c, $tmpl);
#     $c->stash->{template} = 'test.xt';
# }
#
sub test_expose_methods
    : Local
{
    my ($self, $c) = @_;

    $c->stash(exposed => 'ok');
    
    my $return = $c
      ->view('Xslate::ExposeMethods')
      ->render($c, \'hello <: $abc() :> world <: $def("arg") :>');

    $c->response->body($return);
}

sub test_expose_methods_coerced
    : Local
{
    my ($self, $c) = @_;

    $c->stash(exposed => 'ok');
    
    my $return = $c
      ->view('Xslate::ExposeMethodsCoerced')
      ->render($c, \'hello <: $abc() :> world <: $def("arg") :>');

    $c->response->body($return);
}

sub test_header_footer
    : Local
{
    my ($self, $c) = @_;

    my $return = $c->view('Xslate::HeaderFooter')->render($c, \'content!');
    $c->response->body($return);
}

sub end 
    :Private
{
    my ($self, $c) = @_;

    return 1 if $c->response->status =~ /^3\d\d$/;
    return 1 if $c->response->body;

    my $view = 'View::Xslate::' . ($c->request->param('view') || $c->config->{default_view});
    $c->forward($view);
}


__PACKAGE__->meta->make_immutable();

1;


t/lib/TestApp/View/Xslate/Appconfig.pm
package TestApp::View::Xslate::Appconfig;

use strict;
use base 'Catalyst::View::Xslate';

__PACKAGE__->config(
  verbose => 0
);

1;


lib/Catalyst/View/Xslate.pm
package Catalyst::View::Xslate;
use Moose;
use Moose::Util::TypeConstraints qw(coerce from where via subtype);
use Encode;
use Text::Xslate;
use namespace::autoclean;
use Scalar::Util qw/blessed weaken/;
use File::Find ();

our $VERSION = '0.00018';

extends 'Catalyst::View';

with 'Catalyst::Component::ApplicationAttribute';

has catalyst_var => (
    is => 'rw',
    isa => 'Str',
    default => 'c'
);

has template_extension => (
    is => 'rw',
    isa => 'Str',
    default => '.tx'
);

has content_charset => (
    is => 'rw',
    isa => 'Str',
    default => 'UTF-8'
);

has encode_body => (
    is => 'rw',
    isa => 'Bool',
    default => 1,
);

my $clearer = sub { $_[0]->clear_xslate };

has path => (
    is => 'rw',
    isa => 'ArrayRef',
    trigger => $clearer,
    lazy => 1, builder => '_build_path',
);

sub _build_path { return [ shift->_app->path_to('root') ] }

has cache_dir => (
    is => 'rw',
    isa => 'Str',
    trigger => $clearer,
);

has cache => (
    is => 'rw',
    isa => 'Int',
    default => 1,
    trigger => $clearer,
);

has function => (
    is => 'rw',
    isa => 'HashRef',
    default => sub { +{} },
    trigger => $clearer,
);

has footer => (
    is => 'rw',
    isa => 'ArrayRef',
    default => sub { +[] },
    trigger => $clearer
);

has header => (
    is => 'rw',
    isa => 'ArrayRef',
    default => sub { +[] },
    trigger => $clearer
);

has module => (
    is => 'rw',
    isa => 'ArrayRef',
    default => sub { +[] },
    trigger => $clearer,
);

has input_layer => (
    is => 'rw',
    isa => 'Str',
    trigger => $clearer,
);

has syntax => (
    is => 'rw',
    isa => 'Str',
    trigger => $clearer,
);

has escape => (
    is => 'rw',
    isa => 'Str',
    trigger => $clearer,
);

has type => (
    is => 'rw',
    isa => 'Str',
    trigger => $clearer,
);

has suffix => (
    is => 'rw',
    isa => 'Str',
    trigger => $clearer,
    default => '.tx',
);

has verbose => (
    is => 'rw',
    isa => 'Int',
    trigger => $clearer,
);

has xslate => (
    is => 'rw',
    isa => 'Text::Xslate',
    clearer => 'clear_xslate',
    lazy => 1, builder => '_build_xslate',
);

has [qw/line_start tag_start tag_end/] => (is=>'rw', isa=>'Str');
has [qw/warn_handler die_handler pre_process_handler/] => (is=>'rw', isa=>'CodeRef');


my $expose_methods_tc = subtype 'HashRef', where { $_ };
coerce $expose_methods_tc,
  from 'ArrayRef',
  via {
    my %values = map { $_ => $_ } @$_;
    return \%values;
  };

has expose_methods => (
    is => 'ro',
    isa => $expose_methods_tc,
    predicate => 'has_expose_methods',
    coerce => 1,
);

has preload => (
    is => 'ro',
    isa => 'Bool',
    default => 1,
);

sub _build_xslate {
    my $self = shift;

    my $app = $self->_app;
    my $name = $app;
    $name =~ s/::/_/g;

    my %args = (
        path      => $self->path,
        cache_dir => $self->cache_dir || File::Spec->catdir(File::Spec->tmpdir, $name),
        map { ($_ => $self->$_) }
            qw( cache footer function header module )
    );

    # optional stuff
    foreach my $field ( qw( input_layer syntax escape verbose suffix type line_start tag_start tag_end warn_handler die_handler pre_process_handler) ) {
        if (defined(my $value = $self->$field)) {
            $args{$field} = $value;
        }
    }

    return Text::Xslate->new(%args);
}

sub BUILD {
    my $self = shift;
    if ($self->preload) {
        $self->preload_templates();
    }
}

sub preload_templates {
    my $self = shift;
    my ( $paths, $suffix ) = ( $self->path, $self->suffix );
    my $xslate = $self->xslate;
    foreach my $path (@$paths) {
        File::Find::find( sub {
            if (/\Q$suffix\E$/) {
                my $file = $File::Find::name;
                $file =~ s/\Q$path\E .//xsm;
                $xslate->load_file($file);
            }
        }, $path);
    }
}

sub process {
    my ($self, $c) = @_;

    my $stash = $c->stash;
    my $template = $stash->{template} || $c->action . $self->template_extension;

    if (! defined $template) {
        $c->log->debug('No template specified for rendering') if $c->debug;
        return 0;
    }

    my $output = eval {
        $self->render( $c, $template, $stash );
    };
    if (my $err = $@) {
        return $self->_rendering_error($c, $err);
    }

    my $res = $c->response;
    if (! $res->content_type) {
        $res->content_type('text/html; charset=' . $self->content_charset);
    }

    if ( $self->encode_body ) {
        $res->body( encode($self->content_charset, $output) );
    }
    else {
        $res->body( $output );
    }

    return 1;
}

sub build_exposed_method {
    my ( $self, $ctx, $code ) = @_;
    my $weak_ctx = $ctx;
    weaken $weak_ctx;

    return sub { $self->$code($weak_ctx, @_) };
}

sub render {
    my ($self, $c, $template, $vars) = @_;

    $vars = $vars ? $vars : $c->stash;

    if ($self->has_expose_methods) {
        foreach my $exposed_method( keys %{$self->expose_methods} ) {
            if(my $code = $self->can( $self->expose_methods->{$exposed_method} )) {
                $vars->{$exposed_method} = $self->build_exposed_method($c, $code);
            } else {
                Catalyst::Exception->throw( "$exposed_method not found in Xslate view" );
            }

        }
    }

    local $vars->{ $self->catalyst_var } =
        $vars->{ $self->catalyst_var } || $c;

    if(ref $template eq 'SCALAR') {
        return $self->xslate->render_string( $$template, $vars );
    } else {
        return $self->xslate->render($template, $vars );
    }
}

sub _rendering_error {
    my ($self, $c, $err) = @_;
    my $error = qq/Couldn't render template "$err"/;
    $c->log->error($error);
    $c->error($error);
    return 0;
}

__PACKAGE__->meta->make_immutable();

1;

__END__

=head1 NAME

Catalyst::View::Xslate - Text::Xslate View Class

=head1 SYNOPSIS

    package MyApp::View::Xslate;
    use Moose;
    extends 'Catalyst::View::Xslate';

    1;

=head1 VIEW CONFIGURATION

You may specify the following configuration items in from your config file
or directly on the view object.

=head2 catalyst_var

The name used to refer to the Catalyst app object in the template

=head2 template_extension

The suffix used to auto generate the template name from the action name
(when you do not explicitly specify the template filename);

Do not confuse this with the C<suffix> option, which is passed directly to
the Text::Xslate object instance. This option works on the filename used
for the initial request, while C<suffix> controls what C<cascade> and
C<include> directives do inside Text::Xslate.

=head2 content_charset

The charset used to output the response body. The value defaults to 'UTF-8'.

=head2 encode_body

By default, output will be encoded to C<content_charset>.
You can set it to 0 to disable this behavior.
(you need to do this if you're using C<Catalyst::Plugin::Unicode::Encoding>)

B<NOTE> Starting with L<Catalyst> version 5.90080 Catalyst will automatically
encode to UTF8 any text like body responses.  You should either turn off the
body encoding step in this view using this attribute OR disable this feature
in the application (your subclass of Catalyst.pm).

    MyApp->config(encoding => undef);

Failure to do so will result in double encoding.

=head2 Text::Xslate CONFIGURATION

The following parameters are passed to the Text::Xslate constructor.
When reset during the life cyle of the Catalyst app, these parameters will
cause the previously created underlying Text::Xslate object to be cleared

=head2 path

=head2 cache_dir

=head2 cache

=head2 header

=head2 escape

=head2 type

=head2 footer

=head2 function

=head2 input_layer

=head2 module

=head2 syntax

=head2 verbose

=head2 line_start

=head2 tag_start

=head2 tag_end

=head2 warn_handler

=head2 die_handler

=head2 pre_process_handler

=head2 suffix


Use this to enable TT2 compatible variable methods via Text::Xslate::Bridge::TT2 or Text::Xslate::Bridge::TT2Like

    package MyApp::View::Xslate;
    use Moose;
    extends 'Catalyst::View::Xslate';

    has '+module' => (
        default => sub { [ 'Text::Xslate::Bridge::TT2Like' ] }
    );

=head1 preload

Boolean flag indicating if templates should be preloaded. By default this is enabled.

Preloading templates will basically cutdown the cost of template compilation for the first hit.

=head2 expose_methods

Use this option to specify methods from the View object to be exposed in the
template. For example, if you have the following View:

    package MyApp::View::Xslate;
    use Moose;
    extends 'Catalyst::View::Xslate';

    sub foo {
        my ( $self, $c, @args ) = @_;
        return ...; # do something with $self, $c, @args
    }

then by setting expose_methods, you will be able to use $foo() as a function in
the template:

    <: $foo("a", "b", "c") # calls $view->foo( $c, "a", "b", "c" ) :>

C<expose_methods> takes either a list of method names to expose, or a hash reference, in order to alias it differently in the template.

    MyApp::View::Xslate->new(
        # exposes foo(), bar(), baz() in the template
        expose_methods => [ qw(foo bar baz) ]
    );

    MyApp::View::Xslate->new(
        # exposes $foo_alias(), $bar_alias(), $baz_alias() in the template,
        # but they will in turn call foo(), bar(), baz(), on the view object.
        expose_methods => {
            foo => "foo_alias",
            bar => "bar_alias",
            baz => "baz_alias",
        }
    );

NOTE: you can mangle the process of building the exposed methods, see C<build_exposed_method>.

=head1 METHODS

=head1 C<$view->process($c)>

Called by Catalyst.

=head2 C<$view->render($c, $template, \%vars)>

Renders the given C<$template> using variables \%vars.

C<$template> can be a template file name, or a scalar reference to a template
string.

    $view->render($c, "/path/to/a/template.tx", \%vars );

    $view->render($c, \'This is a xslate template!', \%vars );

=head2 C<$view->preload_templates>

Preloads templates in $view-E<gt>path.

=head2 C<$view->build_exposed_method>

Hook point for mangling the building process of exposed methods.

=head1 AUTHOR

Copyright (c) 2010 Daisuke Maki C<< <daisuke@endeworks.jp> >>

=head1 LICENSE 

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See http://www.perl.com/perl/misc/Artistic.html

=cut
